spec/index.bs

```diff
@@ -1277,10 +1282,12 @@ A [=device=] also has the following [=content timeline property=]:
 
     - Set |device|.{{device/[[adapter]]}} to |adapter|.
 
-    - Set |device|.{{device/[[features]]}} to the [=ordered set|set=] of values in
+    - Set |device|.{{device/[[features]]}} to the [=ordered set|set=] of features provided
```

Comment:
Just trying to think through this

Let's say `'float32-filterable'` is a default feature of `'fl2'`. I do

```js
adapter = await navigator.gpu.requestAdapter({ minFeatureLevel: 'fl1' });
```

UA returns an `'fl2'` adapter (because it supports `'fl2'`). Does `adapter.features` contain `'float32-filterable'`?). I'd think yes because my `fl1` app wants to enable it but it just so happens that enabling it is a no-op because it's defaulted to be enabled in `'fl2`'. That means if I do this

```js
device = await adapter.requestDevice();
```

then `device.features` will have `float32-filterable` in it even though I didn't explicitly request it because the UA returned an `'fl2'` device

Note: I don't think this is an issue, I just want to walk through it. I guess in particular, an `'fl2'` adapter has to advertise as available, all `'fl1'` features that default to being enabled in `'fl2'`

Reply:
I would definitely say that `float32-filterable` should show up in `device.features` in both of those scenarios. In my mind it's similar to the limits, in that we wouldn't continue to report the base limits in the device if the feature level included a higher limit.

Reply:
yes, but it also needs to show up in `adapter.features` even though it's a default feature of `'fl2'` (in this example). Otherwise my `fl1` code stops working.

```js
adapter = await navigator.gpu.requestAdapter({ minFeatureLevel: 'fl1' }); // returns fl2 adapter
canUseF32Filter = adapter.features.has('float32-filterable');
device = await adapter.requestDevice({
  requiredFeatures: canUseF32Filterable ? ['float32-filterable'] : [],
});

...

if (canUseF32Filter) { .... }
```

To put it another way: Again, assume `'fl2'` includes `'float32-filterable'`

* If I ask for `'fl1'` and get `'fl2'`, because I asked for `'fl1'`, `'float32-filterable'` must appear in `adapter.features` and, even if I don't request `'float32-filterable'` it will appear in `device.features`

* If I ask for `'fl2'`, `'float32-filterable`' does or does not appear in `adapter.features` nor `device.features`?

   * argument for "does not appear"
    
     because it's part of `'fl2'` directly there's nothing to check for. Any app 
     requesting `'fl2'` already knows this.

   * argument for "does appear"

     because core/libraries need to know what features exist regardless and not
     suddenly stop working because a feature no longer shows up in `(adapter|device).features`.

This seems like what the spec really wants to say is that requesting a feature level just pre-selects features and limits as though the user requested them manually.

For `requestAdapter` it includes any features required by the requested feature level (meaning they must show up in `adapter.features`)

For `requestDevice`, it's as though it inserted the features required for this and previous feature levels into `requiredFeatures` 

```js
GPUAdapter.prototype.requestDevice = function(desc) {
  const adapter = this;
  const updatedDescriptor = {
    requiredLimits: desc?.requiredLimits ?? {},
    requiredFeatures: [
      ...(desc.requiredFeatures ?? [])
      ...allFeaturesFromFeatureLevelAndBelow(adapter.featureLevel),
    ],
  };
  return adapter.realRequestDevice(updatedAdapter);
});
```

Reply:
I believe we're in aggressive agreement here. 😁 

The only somewhat strange aspect to me is that there isn't anything described so far in the adapter that says "these are the limits and features you'll be getting", which means the feature level kinda disappears in between requestAdapter and requestDevice. That's probably not an issue? It just feels a little weird.